pkg/exposer/csi_snapshot.go

```diff
@@ -0,0 +1,386 @@
+		if apierrors.IsNotFound(err) {
+			curLog.WithField("backup pod", backupPodName).Errorf("Backup pod is not running in the current node %s", exposeWaitParam.NodeName)
+			return nil, nil
+		} else {
+			return nil, errors.Wrapf(err, "error to get backup pod %s", backupPodName)
+		}
+	}
+
+	curLog.WithField("pod", pod.Name).Infof("Backup pod is in running state in node %s", pod.Spec.NodeName)
+
+	_, err = kube.WaitPVCBound(ctx, e.kubeClient.CoreV1(), e.kubeClient.CoreV1(), backupPVCName, ownerObject.GetNamespace(), timeout)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error to wait backup PVC bound, %s", backupPVCName)
+	}
+
+	curLog.WithField("backup pvc", backupPVCName).Info("Backup PVC is bound")
+
+	return &ExposeResult{ByPod: ExposeByPod{HostingPod: pod, PVC: backupPVCName}}, nil
+}
+
+func (e *csiSnapshotExposer) CleanUp(ctx context.Context, ownerObject *unstructured.Unstructured, vsName string, sourceNamespace string) {
+	backupPodName := ownerObject.GetName()
+	backupPVCName := ownerObject.GetName()
+	backupVSName := ownerObject.GetName()
+
+	kube.DeletePodIfAny(ctx, e.kubeClient.CoreV1(), backupPodName, ownerObject.GetNamespace(), e.log)
+	kube.DeletePVCIfAny(ctx, e.kubeClient.CoreV1(), backupPVCName, ownerObject.GetNamespace(), e.log)
+	csi.DeleteVolumeSnapshotIfAny(ctx, e.csiSnapshotClient, backupVSName, ownerObject.GetNamespace(), e.log)
+	csi.DeleteVolumeSnapshotIfAny(ctx, e.csiSnapshotClient, vsName, sourceNamespace, e.log)
+}
+
+func getVolumeModeByAccessMode(accessMode string) (corev1.PersistentVolumeMode, error) {
+	if accessMode == AccessModeFileSystem {
+		return corev1.PersistentVolumeFilesystem, nil
+	} else {
+		return "", errors.Errorf("unsupported access mode %s", accessMode)
+	}
+}
+
+func (e *csiSnapshotExposer) createBackupVS(ctx context.Context, ownerObject *unstructured.Unstructured, snapshotVS *snapshotv1api.VolumeSnapshot) (*snapshotv1api.VolumeSnapshot, error) {
+	backupVSName := ownerObject.GetName()
+	backupVSCName := ownerObject.GetName()
+
+	vs := &snapshotv1api.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backupVSName,
+			Namespace: ownerObject.GetNamespace(),
+			// Don't add ownerReference to SnapshotBackup.
+			// The backupPVC should be deleted before backupVS, otherwise, the deletion of backupVS will fail since
+			// backupPVC has its dataSource referring to it
+		},
+		Spec: snapshotv1api.VolumeSnapshotSpec{
+			Source: snapshotv1api.VolumeSnapshotSource{
+				VolumeSnapshotContentName: &backupVSCName,
+			},
+			VolumeSnapshotClassName: snapshotVS.Spec.VolumeSnapshotClassName,
+		},
+	}
+
+	return e.csiSnapshotClient.VolumeSnapshots(vs.Namespace).Create(ctx, vs, metav1.CreateOptions{})
+}
+
+func (e *csiSnapshotExposer) createBackupVSC(ctx context.Context, ownerObject *unstructured.Unstructured, snapshotVSC *snapshotv1api.VolumeSnapshotContent, vs *snapshotv1api.VolumeSnapshot) (*snapshotv1api.VolumeSnapshotContent, error) {
+	backupVSCName := ownerObject.GetName()
```

Comment:
It seems the VSC name using the `ownerObject` name is not right, because VSC is not a namespaced-scope resource.
If two `ownerObject` from different namespaces have the same name, this will cause conflict.

Reply:
Actually, this doesn't matter. The ownerObjects passed here are guaranteed to be in `velero` namespace. Since they are in the same namespace, their name must not be conflicted.